tests/test_em27_opus.py

```diff
@@ -64,3 +65,77 @@ def test_make_request_error(requests_mock) -> None:
         )
 
         request_error_mock.emit.assert_called_once_with(error)
+
+
+def _format_td(name: str, value: Any) -> str:
+    if value is None:
+        return ""
+    return f'<td id="{name}">{str(value)}</td>'
+
+
+def _get_opus_html(
+    status: Optional[int],
+    text: Optional[str],
+    errcode: Optional[int],
+    errtext: Optional[str],
+) -> str:
+    return f"""
+    <html>
+        <body>
+            <table>
+                <tr>
+                    {_format_td("STATUS", status)}
+                    {_format_td("TEXT", text)}
+                    {_format_td("ERRCODE", errcode)}
+                    {_format_td("ERRTEXT", errtext)}
+                </tr>
+            </table>
+        </body>
+    </html>
+    """
+
+
+@pytest.mark.parametrize(
+    "http_status_code,status,text,errcode,errtext",
+    (
+        (http_status, status, text, errcode, errtext)
+        for http_status in (200, 403, 404)
+        for status in (None, 0, 1)
+        for text in (None, "", "status text")
+        for errcode in (None, 0, 1)
+        for errtext in (None, "", "error text")
+    ),
+)
+def test_parsing(
```

Comment:
Having parametric testing is really useful if you can provided the straight, expected output as an input to the test, so the logic of the test is also simple. In this case, You have a complex logic of nested conditional statement (in blocks and inline) that, in my view, defeat the purpose of having para metric testing since it does not make the test simpler. 

I would suggest splitting this test into several ones, potentially each parametrized, but much less than this one, such that the test logic involves no or minimal conditionals.

Reply:
Good point. This *felt* wrong when I was writing it actually :laughing:. I'll try breaking it up.